src/dune_rules/melange_rules.ml

```diff
@@ -60,25 +66,116 @@ let build_js ~pkg_name ~module_system ~dst_dir ~cctx m =
+        Obj_dir.make_exe ~dir:emit_stanza_dir ~name:mel.target
+      in
+      let open Memo.O in
+      let* cctx =
+        let flags = Ocaml_flags.empty in
+        let vimpl =
+          (* original impl in Lib_rules uses Virtual_rules.impl, will this break with virtual libs? *)
+          None
+        in
+        let modules = Vimpl.impl_modules vimpl (Modules.singleton m) in
+        let requires_compile = Lib.Compile.direct_requires compile_info in
+        let requires_link = Lib.Compile.requires_link compile_info in
+        let js_of_ocaml = None in
+        (* modes and pp are not passed, not sure if this will cause issues *)
+        Compilation_context.create () ~super_context:sctx ~expander ~scope
+          ~obj_dir ~modules ~flags ~requires_compile ~requires_link
+          ~opaque:Inherit_from_settings ~js_of_ocaml ~package:mel.package ?vimpl
+          ~modes:
+            { ocaml = { byte = None; native = None }
+            ; melange = Some (Requested Loc.none)
+            }
+      in
+      let pkg_name = Option.map mel.package ~f:Package.name in
+      let dst_dir = Path.Build.relative emit_stanza_dir mel.target in
+      let loc = mel.loc in
+      let lib_deps_js_includes =
+        js_includes ~emit_stanza_dir ~target:mel.target ~requires_link ~scope
+      in
+      let build_dir = (Super_context.context sctx).build_dir in
+      (* TODO: build_all should go outside Memo.List.iter, or maybe call a version of build_cm but without cctx *)
+      let* () = Module_compilation.build_all cctx in
+      build_js ~loc ~pkg_name ~module_system:mel.module_system ~dst_dir ~obj_dir
+        ~sctx ~build_dir ~lib_deps_js_includes m)
+
+let add_rules_for_libraries ~scope ~emit_stanza_dir ~sctx ~requires_link
+    (mel : Melange_stanzas.Emit.t) =
+  Memo.List.iter requires_link ~f:(fun lib ->
+      let open Memo.O in
+      let lib_name = Lib.name lib in
+      let* lib, lib_compile_info =
+        Lib.DB.get_compile_info (Scope.libs scope) lib_name
+      in
+      let lib = Lib.Local.of_lib_exn lib in
+      let info = Lib.Local.info lib in
+      let lib_dir = Lib_info.src_dir info in
+      let obj_dir = Lib_info.obj_dir info in
+      let () =
+        if not (Path.Build.is_descendant lib_dir ~of_:emit_stanza_dir) then
+          User_error.raise
+            [ Pp.textf
+                "The library %s is used by a melange.emit stanza but the \
+                 library folder %s is not a descendant of the stanza folder %s"
+                (Lib_name.to_string (Lib_info.name info))
+                (Path.Build.to_string lib_dir)
+                (Path.Build.to_string emit_stanza_dir)
+            ]
+      in
+      let dst_dir =
+        Melange.lib_output_dir ~emit_stanza_dir ~lib_dir ~target:mel.target
+      in
+      let modules_group =
+        Dir_contents.get sctx ~dir:lib_dir
+        >>= Dir_contents.ocaml
+        >>| Ml_sources.modules ~for_:(Library lib_name)
```

Comment:
I think this is not the right way to get the modules from a library, I believe it only checks one folder, and does not take things like `modules` field or `include_subdirs unqualified` into account. Is there another way?